File: java/src/com/test/sorting/Shell.java
package com.test.sorting;


public class Shell extends Sorter
{
    public Shell(int[] arr)
    {
        super(arr);
    }

    @Override
    public void sort()
    {
        int N = arr.length;
        int h = 1;
        while (h < N/3) {
            h = 3*h + 1;
        }
        while (h >= 1) {
            for (int i = h; i < N; i++) {
                for (int j = i; j >= h && less(arr[j], arr[j-h]); j-=h) {
                    exch(j, j-h);
                }
            }

            h = h/3;
        }
    }
}

File: js/src/sorting/Shell.js
import Sorter from './Sorter';

class Shell extends Sorter {
    constructor(arr = []) {
        super(arr);
    }

    sort() {
        let N = this.arr.length;
        let h = 1;
        while (h < N/3 ) {
            h = Math.floor(3*h + 1);
        }

        while (h >= 1) {
            for (let i = h; i < N; i++) {
                for (let j = i; j >= h && this.less(this.arr[j], this.arr[j-h]); j -= h) {
                    this.exch(j, j-h);
                }
            }

            h = Math.floor(h/3);
        }
    }
}

module.exports = Shell;